x-pack/plugins/endpoint/public/applications/endpoint/view/hooks/use_navigate_by_router_event_handler.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { MouseEventHandler, useCallback } from 'react';
+import { useHistory } from 'react-router-dom';
+import { LocationDescriptorObject } from 'history';
+
+type EventHandlerCallback = MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>;
+
+/**
+ * Provides an event handler that can be used with (for example) `onClick` props to prevent the
+ * event's default behaviour and instead navigate to to a route via the Router
+ *
+ * @param routeTo
+ * @param onClick
+ */
+export const useNavigateByRouterEventHandler = (
+  routeTo: string | [string, unknown] | LocationDescriptorObject<unknown>, // Cover the calling signature of `history.push()`
+
+  /** Additional onClick callback */
+  onClick?: EventHandlerCallback
```

Comment:
Maybe you could add a type parameter to this hook like: `TargetElement extends HTMLButtonElement | HTMLAnchorElement`
and then define `onClick` like: `onClick?: MouseEventHandler<TargetElement>` and then the return type of the hook could be `MouseEventHandler<TargetElement>`

I think this would allow `onClick` to have button or link specific logic (if needed)

Reply:
🤔 isn't that the same that I already have defined above in the `EventHandlerCallback` type?

Reply:
I don't see the difference here with what @paul-tavares did vs @oatkiller suggestion.  Maybe we could reuse `TargetElement` if we defined it outside of the the generic types for `MouseEventHandler`.  But we could do that later if needed